master/static/migrations/20230509102913_add-generic-metrics-partitioned-table.tx.up.sql

```diff
@@ -0,0 +1,132 @@
+ "steps_archived" btree (archived)
+ Foreign-key constraints:
+ "steps_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(id)
+ 
+ determined> \d raw_validations;
+ +---------------+--------------------------+--------------------------------------------+
+ | Column        | Type                     | Modifiers                                  |
+ |---------------+--------------------------+--------------------------------------------|
+ | id            | integer                  |  not null generated by default as identity |
+ | trial_id      | integer                  |  not null                                  |
+ | end_time      | timestamp with time zone |                                            |
+ | metrics       | jsonb                    |                                            |
+ | total_batches | integer                  |  not null default 0                        |
+ | trial_run_id  | integer                  |  not null default 0                        |
+ | archived      | boolean                  |  not null default false                    |
+ +---------------+--------------------------+--------------------------------------------+
+ Indexes:
+ "validations_pkey" PRIMARY KEY, btree (id)
+ "validations_trial_id_total_batches_run_id_unique" UNIQUE, btree (trial_id, total_batches, trial_run_id)
+ "ix_validations_trial_id" btree (trial_id)
+ "validations_archived" btree (archived)
+ Referenced by:
+ TABLE "trials" CONSTRAINT "trials_latest_validation_id_fkey" FOREIGN KEY (latest_validation_id) REFERENCES raw_validations(id) ON DELETE SET NULL
+ Triggers:
+ autoupdate_exp_validation_metrics_name AFTER INSERT ON raw_validations FOR EACH ROW EXECUTE PROCEDURE autoupdate_exp_validation_metrics_name()
+ */
+-- fix missing foreign key constraint
+ALTER TABLE raw_validations
+ADD CONSTRAINT validations_trial_id_fkey FOREIGN KEY (trial_id) REFERENCES trials(id);
+CREATE TYPE metric_partition_type AS ENUM ('validation', 'training', 'generic');
+ALTER TABLE raw_validations
+ADD COLUMN partition_type metric_partition_type NOT NULL DEFAULT 'validation';
+ALTER INDEX validations_trial_id_total_batches_run_id_unique
+RENAME TO validations_trial_id_total_batches_run_id_unique_old;
+CREATE UNIQUE INDEX validations_trial_id_total_batches_run_id_type_unique ON raw_validations (
+    trial_id,
+    total_batches,
+    trial_run_id,
+    partition_type
+);
+DROP INDEX validations_trial_id_total_batches_run_id_unique_old;
+ALTER TABLE raw_steps
+ADD COLUMN partition_type metric_partition_type NOT NULL DEFAULT 'training';
+ALTER INDEX steps_trial_id_total_batches_run_id_unique
+RENAME TO steps_trial_id_total_batches_run_id_unique_old;
+CREATE UNIQUE INDEX steps_trial_id_total_batches_run_id_type_unique ON raw_steps (
+    trial_id,
+    total_batches,
+    trial_run_id,
+    partition_type
+);
+DROP INDEX steps_trial_id_total_batches_run_id_unique_old;
+ALTER TABLE raw_validations
+ADD COLUMN custom_type text;
```

Comment:
- maybe call it `metric_type`? that's what it's currently called in the APIs
- also default it to training/validations, depending on the table, in case if we get rid of "partition_type" when we upgrade to the newer postgres?

Reply:
- re name:  I saw a few references of metric_types in summary metrics which were about the metric value type so I ended up changing these. 
- defaulting: the way the db handlers and schema is set up right now we could be pushing different values there and I'm thinking it'd leave us with more options if we don't default them for now

Reply:
write performance is mainly the same.
working on testing migration and read perf